File: swift/OpenMedKit/Sources/OpenMedKit/EntityPrediction.swift
import Foundation

/// A single entity predicted by the NER pipeline.
public struct EntityPrediction: Codable, Equatable, Sendable {
    /// The entity type label (e.g., "first_name", "date_of_birth", "ssn").
    public let label: String

    /// The text span matched by this entity.
    public let text: String

    /// Model confidence score (0.0 – 1.0).
    public let confidence: Float

    /// Inclusive Unicode scalar (code point) offset in the original text.
    public let start: Int

    /// Exclusive Unicode scalar (code point) offset in the original text.
    public let end: Int

    /// Python-compatible entity type used by de-identification exports.
    public var entityType: String {
        label
    }

    public init(label: String, text: String, confidence: Float, start: Int, end: Int) {
        self.label = label
        self.text = text
        self.confidence = confidence
        self.start = start
        self.end = end
    }

    /// Return the Swift range represented by this entity's scalar offsets.
    public func range(in source: String) -> Range<String.Index>? {
        PostProcessing.scalarRange(in: source, start: start, end: end)
    }

    /// Return a copy whose offsets enclose complete grapheme clusters.
    ///
    /// The copied `text` is sliced from `source` so it always agrees with the
    /// adjusted offsets.
    public func snappedToGraphemeBoundaries(
        in source: String
    ) -> EntityPrediction? {
        let snapped = PostProcessing.snapScalarSpanToGraphemeBoundaries(
            start: start,
            end: end,
            in: source
        )
        guard
            PostProcessing.scalarRange(
                in: source,
                start: snapped.start,
                end: snapped.end
            ) != nil
        else {
            return nil
        }
        return EntityPrediction(
            label: label,
            text: PostProcessing.scalarSubstring(
                source,
                start: snapped.start,
                end: snapped.end
            ),
            confidence: confidence,
            start: snapped.start,
            end: snapped.end
        )
    }
}

extension EntityPrediction: CustomStringConvertible {
    public var description: String {
        "[\(label)] \"\(text)\" (\(start):\(end)) conf=\(String(format: "%.2f", confidence))"
    }
}

File: android/openmedkit/src/main/kotlin/com/openmed/openmedkit/EntityPrediction.kt
package com.openmed.openmedkit

import com.openmed.openmedkit.segmentation.IcuTextSegmenter
import java.math.BigDecimal
import java.math.RoundingMode

/**
 * A single entity predicted by the OpenMedKit token-classification pipeline.
 *
 * Mirrors the Swift `OpenMedKit.EntityPrediction` value type and the Python
 * `OpenMedSpan` conventions: [start] and [end] are half-open Unicode scalar
 * (code point) offsets into the exact original text. They are never Kotlin
 * UTF-16 indices.
 */
data class EntityPrediction(
    val label: String,
    val text: String,
    val confidence: Float,
    val start: Int,
    val end: Int,
) {
    /** Python-compatible entity type used by de-identification exports. */
    val entityType: String
        get() = label

    /**
     * Return the native Kotlin UTF-16 range represented by this entity.
     */
    fun utf16SpanIn(source: String): Utf16Span? {
        val scalarLength = UnicodeOffsetContract.scalarLength(source)
        if (start !in 0..scalarLength || end !in start..scalarLength) {
            return null
        }
        return UnicodeOffsetContract.utf16Span(source, start, end)
    }

    /**
     * Return a copy whose scalar offsets enclose complete grapheme clusters.
     *
     * The copied [text] is sliced from [source] after converting the snapped
     * scalar coordinates to Kotlin's native UTF-16 indices.
     */
    fun snappedToGraphemeBoundaries(
        source: String,
        segmenter: IcuTextSegmenter = IcuTextSegmenter(),
    ): EntityPrediction {
        val snapped = segmenter.snapScalarSpan(source, start, end)
        return copy(
            text = UnicodeOffsetContract.substring(
                source,
                snapped.start,
                snapped.end,
            ),
            start = snapped.start,
            end = snapped.end,
        )
    }

    /**
     * Human-readable description matching the Swift `EntityPrediction`
     * format `[label] "text" (start:end) conf=0.00`.
     *
     * The confidence is rendered with two decimals using half-even
     * ("banker's") rounding on a period decimal separator, matching Swift's
     * `String(format: "%.2f", confidence)` (C `printf`) exactly, including
     * ties on exactly-representable halves such as `0.125 -> 0.12`.
     * Non-finite confidences are rendered without rounding so this never
     * throws.
     */
    override fun toString(): String {
        val conf =
            if (confidence.isFinite()) {
                BigDecimal(confidence.toDouble())
                    .setScale(2, RoundingMode.HALF_EVEN)
                    .toPlainString()
            } else {
                confidence.toString()
            }
        return "[$label] \"$text\" ($start:$end) conf=$conf"
    }
}
